Add basePath documentation (#81)
flightcontrolhq/website: motif-docs/pages/guides/config/monorepos.mdx

---
heading: Configuration
title: Monorepo Setup
---

## Configuration File

You can place the config file at the root of your repository and use it to configure each Flightcontrol service. See below some guidance on how to configure them.

## Using `buildType: nixpacks`

There are different types of monorepos and tooling around it. Your configuration will depend on those.
We cover a few cases below.

### Independent packages in a repo

If you don't use any sort of monorepo tooling at the root of your repo, then likely your configuration will only need to indicate a `basePath`, like illustrated in the contrived config example below.

> If for some reason `nixpacks` cannot identify the `[install|build|start]Command`, you'll then need to manually add it.

Reviewer: it won't, so they almost certainly need to add this

      },
      "services": [
        {
          "id": "api",
          "name": "Backend API",
          "type": "fargate",
          "buildType": "nixpacks",
          "basePath": "./packages/api",
          "healthCheckPath": "/graphql/health"
        },
        {
          "id": "frontend",
          "name": "Frontend Web",
          "type": "static",
          "singlePageApp": true,
          "basePath": "./packages/frontend",
          "outputDirectory": "web/dist",
        }
      ]
    }
  ]
}
```

### Packages managed with monorepo tooling

For JS monorepo that use `npm` and/or require installation at the root, you'll likely need to manually specify the install/build command AND account for the right path by `cd`'ing to the corresponding package, like so:

```json
{
  "environments": [
    {
      ...
      "services": [
        {
          "id": "api",
          "buildType": "nixpacks",

          "installCommand": "npm i && cd packages/api && npm i"
          "buildCommand": "cd packages/api && npm run build",
          "startCommand": "cd packages/api && npm run start",

          "basePath": "./",       // This is the default, no need to include it.

>                                      Left for illustration purpose

Reviewer: ```suggestion
          "installCommand": "cd .. && npm ci"
          "basePath": "packages/api", 
```

### Independent packages in a repo

> If you don't use any sort of monorepo tooling at the root of your repo, then you should indicate a `basePath` alongside the `[install|build|start]` commands, like illustrated in the contrived config example below.

Reviewer: ```suggestion
Most users should define a `basePath` alongside the `[install|build|start]` commands, as illustrated below.
```

Author: Fixes as we discussed

      envVariables": {
        "NIXPACKS_CONFIG_FILE": "./apps/frontend/nixpacks.toml"
      }
      ...
    },
    {
      id: "api",
      buildType: "nixpacks",
      envVariables": {
        "NIXPACKS_CONFIG_FILE": "./packages/api/nixpacks.toml"
      }
      ...
    },
  ],

> }

Reviewer: Instead of using `NIXPACKS_CONFIG_FILE`, they should define `basePath` to the service, and then add `nixpacks.toml` in that basePath location.
